test-data/unit/check-attr.test

```diff
@@ -0,0 +1,444 @@
+
+A([B()])
+[builtins fixtures/list.pyi]
+
+[case testAttrsUsingConvert]
+import attr
+
+def convert(s:int) -> str:
+    return 'hello'
+
+@attr.s
+class C:
+    x: str = attr.ib(convert=convert)
+
+# Because of the convert the __init__ takes an int, but the variable is a str.
+reveal_type(C)  # E: Revealed type is 'def (x: builtins.int) -> __main__.C'
+reveal_type(C(15).x)  # E: Revealed type is 'builtins.str'
+[builtins fixtures/list.pyi]
+
+[case testAttrsUsingConverter]
+import attr
+
+def converter(s:int) -> str:
+    return 'hello'
+
+@attr.s
+class C:
+    x: str = attr.ib(converter=converter)
+
+# Because of the converter the __init__ takes an int, but the variable is a str.
+reveal_type(C)  # E: Revealed type is 'def (x: builtins.int) -> __main__.C'
+reveal_type(C(15).x)  # E: Revealed type is 'builtins.str'
+[builtins fixtures/list.pyi]
+
+[case testAttrsUsingConvertAndConverter]
+import attr
+
+def converter(s:int) -> str:
+    return 'hello'
+
+@attr.s
+class C:
+    x: str = attr.ib(converter=converter, convert=converter)  # E: Can't pass both `convert` and `converter`.
+
+[builtins fixtures/list.pyi]
+
+
+[case testAttrsCmpWithSubclasses]
+import attr
+@attr.s
+class A: pass
+@attr.s
+class B: pass
+@attr.s
+class C(A, B): pass
+@attr.s
+class D(A): pass
+
+reveal_type(A.__lt__)  # E: Revealed type is 'def [AT] (self: AT`1, other: AT`1) -> builtins.bool'
+reveal_type(B.__lt__)  # E: Revealed type is 'def [AT] (self: AT`1, other: AT`1) -> builtins.bool'
+reveal_type(C.__lt__)  # E: Revealed type is 'def [AT] (self: AT`1, other: AT`1) -> builtins.bool'
+reveal_type(D.__lt__)  # E: Revealed type is 'def [AT] (self: AT`1, other: AT`1) -> builtins.bool'
+
+A() < A()
+B() < B()
+A() < B() # E: Unsupported operand types for > ("B" and "A")
+
+C() > A()
+C() > B()
+C() > C()
+C() > D() # E: Unsupported operand types for < ("D" and "C")
+
+D() >= A()
+D() >= B()  # E: Unsupported operand types for <= ("B" and "D")
+D() >= C()  # E: Unsupported operand types for <= ("C" and "D")
+D() >= D()
+
+A() <= 1  # E: Unsupported operand types for <= ("A" and "int")
+B() <= 1  # E: Unsupported operand types for <= ("B" and "int")
+C() <= 1  # E: Unsupported operand types for <= ("C" and "int")
+D() <= 1  # E: Unsupported operand types for <= ("D" and "int")
+
+A() == A()
+B() == A()
+C() == A()
+D() == A()
+
+A() == int
+B() == int
+C() == int
+D() == int
+[builtins fixtures/list.pyi]
+
+[case testAttrsComplexSuperclass]
+import attr
+
+import attr
+@attr.s
+class C(object):
+    x: int = attr.ib(default=1)
+    y: int = attr.ib()
+    @y.default
+    def name_does_not_matter(self):
+        return self.x + 1
+
+
+@attr.s
+class A(C):
+    z: int = attr.ib(default=18)
+
+reveal_type(C)  # E: Revealed type is 'def (x: builtins.int =, y: builtins.int =) -> __main__.C'
+reveal_type(A)  # E: Revealed type is 'def (x: builtins.int =, y: builtins.int =, z: builtins.int =) -> __main__.A'
+
+[builtins fixtures/list.pyi]
```

Comment:
A general comment about tests: you are mostly testing situations when everything goes right (i.e. user code is correct). Try to add more tests where a user makes various errors, to be sure that we give a clear error in a correct location.